staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -199,7 +199,7 @@ func (s *specAggregator) updateServiceLocked(name string) error {
 		}
 		group := specInfo.apiService.Spec.Group
 		version := specInfo.apiService.Spec.Version
-		return aggregator.FilterSpecByPathsWithoutSideEffects(result, []string{"/apis/" + group + "/" + version}), etag, nil
+		return aggregator.FilterSpecByPathsWithoutSideEffects(result, []string{"/apis/" + group + "/" + version + "/"}), etag, nil
```

Comment:
do paths in swagger always have trailing slashes? do we have to handle both the exact match and prefix-with-`/`-separator case?

Reply:
I looked at the swagger.json earlier and there were no matches of paths of the format "/apis/group/version" without the trailing slash. I believe this is specifically related to how [version discovery is registered](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/discovery/version.go#L65-L73) and does not apply in general for swagger endpoints.

I believe only version discovery is handled at the `/apis/group/version/` endpoint and all other paths served by an apiserver will have some (`/apis/group/version/<suffixpath>`) suffix path  (whether resource/namespace/etc) that will match the filter with a slash.

Reply:
the two pieces of this are:
1. what do paths in swagger docs look like?
2. how does this filter function work?

If we can have paths in swagger like "/apis/$group/$version" without a trailing slash (I don't know if we can or not), then I don't think this change requiring a trailing slash will work correctly.

It seems like we want FilterSpecByPathsWithoutSideEffects to include exact matches for the input paths or prefixes-with-following-slash. I don't know who is familiar enough with FilterSpecByPathsWithoutSideEffects to review a change like that.

Reply:
1. Paths are of the form `/apis/$group/$version/` with only a trailing slash. In addition,T here are paths of the form `/apis/$group/$version/$resource` and  but those will match the filter and pass without any problems. See https://gist.github.com/Jefftree/3f9e89297951453838b8828d67409bba for the list
2. FilterSpecByPathsWithoutSideEffects [discards all paths that are not matched by the prefix defined in the input](https://github.com/kubernetes/kube-openapi/blob/master/pkg/aggregator/aggregator.go#L72-L77). It will filter out exact matches without the trailing slash but we do not ever publish an OpenAPI without the trailing slash for `/apis/$group/$version/`. I don't think we're trying to change any behavior this function.

Reply:
/assign @deadsk
as reviewer of the original PR in https://github.com/kubernetes/kubernetes/pull/123570 and someone who runs a lot of varied aggregated API servers ... hopefully can provide more context on the desired path (ha) forward here

Reply:
I agree that the safest thing to achieve the goal is to exclude exact matches on prefix and prefix+slash. The required modification looks small.

But the only thing we serve before the slash after version is discovery.

Reply:
> I agree that the safest thing to achieve the goal is to exclude exact matches on prefix and prefix+slash. The required modification looks small.

Agreed, though I just noticed that requires changing kube-openapi and revendoring.

That's tedious in master, but doable. It's significantly more change in release branches after https://github.com/kubernetes/kube-openapi/pull/458 :-/

Reply:
> But the only thing we serve before the slash after version is discovery.

That's true. Given we register discovery with a trailing slash in our libraries, and the only thing that would be excluded from openapi if a rogue aggregated server included an `/apis/$group/$version` path with no trailing slash would be discovery, I can live with this fix as-is for backporting purposes.

I would still like to make FilterSpecByPathsWithoutSideEffects correct at HEAD in kube-openapi when given a path without a trailing slash, and then update kubernetes/kubernetes master to be completely correct in case an aggregated server included a slashless version path.